internal/util/projutil/exec.go

```diff
@@ -0,0 +1,118 @@
+// Copyright 2019 The Operator-SDK Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package projutil
+
+import (
+	"fmt"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strings"
+
+	homedir "github.com/mitchellh/go-homedir"
+)
+
+func ExecCmd(cmd *exec.Cmd) error {
+	cmd.Stdout = os.Stdout
+	cmd.Stderr = os.Stderr
+	if err := cmd.Run(); err != nil {
+		return fmt.Errorf("failed to exec %#v: %v", cmd.Args, err)
+	}
+	return nil
+}
+
+// GoBuildOptions configures "go build" and "go test".
+type GoBuildOptions struct {
+	// BinName is the name of the compiled binary, passed to -o.
+	BinName string
+	// BuildArgs are args passed to "go build", aside from "-o {bin_name}".
+	BuildArgs []string
+	// BuildPath is the path containing main package file.
+	BuildPath string
+	// Env is a list of environment variables to pass to "go build";
+	// exec.Command.Env is set to this value.
+	Env []string
+	// Dir is the dir to run "go build" in; exec.Command.Dir is set to this value.
+	Dir string
+	// GoMod determines whether to set the "-mod=vendor" flag.
+	// If true, "go build" will use modules.
+	// If false, "go build" will not use go modules. This is the default.
+	GoMod bool
+}
+
+const (
+	goBuildCmd = "build"
+	goTestCmd  = "test"
+)
+
+// GoBuild runs "go build" configured with opts.
+func GoBuild(opts GoBuildOptions) error {
+	return goCmd(goBuildCmd, opts)
+}
+
+// GoTest runs "go test" configured with opts.
+func GoTest(opts GoBuildOptions) error {
+	return goCmd(goTestCmd, opts)
+}
+
+// goCmd runs "go {cmd}", where cmd is either "build" or "test".
+func goCmd(cmd string, opts GoBuildOptions) error {
+	bargs := []string{cmd}
+	if opts.BinName != "" {
+		bargs = append(bargs, "-o", opts.BinName)
+	}
+	if cmd == goTestCmd {
+		bargs = append(bargs, opts.BuildPath)
+	}
+	bargs = append(bargs, opts.BuildArgs...)
+	// Modules can be used if either GO111MODULE=on or we're not in $GOPATH/src.
+	if opts.GoMod {
+		inGoPath, err := wdInGoPath()
+		if err != nil {
+			return err
+		}
+		if os.Getenv(GoModEnv) == "on" || !inGoPath {
+			bargs = append(bargs, "-mod=vendor")
+		}
+	}
+	if cmd == goBuildCmd {
+		bargs = append(bargs, opts.BuildPath)
```

Comment:
I'm probably missing something but in Lines 76-78 we're appending the BuildPath for `go test`:
```Go
	if cmd == goTestCmd {
		bargs = append(bargs, opts.BuildPath)
	}
```
And here we do it for `go build` as well. So why not always append this argument at the end for both commands.
```console
$ go help test
usage: go test [build/test flags] [packages] [build/test flags & test binary flags]

$ go help build
usage: go build [-o output] [-i] [build flags] [packages]
```
Is the position of the BuildArgs (before or after the packages) different for `go test` and `go build`?

Reply:
For `go build`, `[packages]` must be last; for `go test` they can be before or after `[build/test flags]` but not after `[test binary flags]`. We must either differentiate between the two `go test` flag sets or just put all flags after `[packages]`; I originally chose the latter. Give your [above comment](https://github.com/operator-framework/operator-sdk/pull/1310#pullrequestreview-230303395), it might be prudent to create a second set of options `GoTestOptions` that embeds `GoCmdOptions` then has an extra `TestBinaryArgs` field for `[test binary flags]`. WDYT?

Reply:
Or we could simply add a `TestBinaryArgs` option that only gets set if `cmd == goTestCmd`.

Reply:
I think having a separate `GoTestOptions` that embeds `GoCmdOptions` would be better. Having a separate `TestBinaryArgs` field only for `GoTest()` wouldn't make `GoCmdOptions` general anymore.